Update the README with better information

RobinKnipe opened the pull request:
...describing some of the important configuration and other settings.

RobinKnipe committed:
Update the README with better information

...describing some of the important configuration and other settings.

aarongill1 reviewed (approved):
LGTM

daniel-ac-martin commented on README.md:
````diff
@@ -1,2 +1,49 @@
+## Running the test
+The performance tests can be started with the following command:
+```shell script
+./entrypoint.sh <path-to-config-file>
````
Is there no config file included that could be referenced?

RobinKnipe replied:
There is now an example/template file. Fixed

daniel-ac-martin commented on README.md:
```diff
@@ -1,2 +1,49 @@
+There are two parts to the configuration:
+ - `./high-load.yml` - the main config file for [Artillery.io], containing:
+   - basic test setup
+   - security details
```
Should this be authentication rather than security?

RobinKnipe replied:
Fixed.

daniel-ac-martin commented on README.md:
```diff
@@ -1,2 +1,49 @@
+ - `./high-load.yml` - the main config file for [Artillery.io], containing:
+   - basic test setup
+   - security details
+   - config for the phases
```
Worth explaining why there are different phases?

RobinKnipe replied:
Well there's currently only a single phase definition. We haven't made use of multiple phases yet.

RobinKnipe replied:
Once we make use of multiple phases, it would probably be worth adding something.

daniel-ac-martin replied:
Oh is the 'ramp-up' not a phase?

RobinKnipe replied:
It could be a phase, but so far Aaron and I haven't found it to be especially useful for our manual testing. If we were to build performance tests into CD, it would likely be useful at that stage.

daniel-ac-martin commented on README.md:
```diff
@@ -1,2 +1,49 @@
+   - config for the phases
+   - payload parameters (the information to match)
+   - scenarios (what requests to send, and how responses are processed)
+ - an extra file supplied to `entry-point.sh`, with security details:
```
Is the hyphen correct?

daniel-ac-martin replied:
Should 'security' be 'authentication'?

RobinKnipe replied:
Both fixed.

daniel-ac-martin commented on README.md:
```diff
@@ -1,2 +1,49 @@
+Each scenario defines a type of dummy user action which Artillery will mimic
+during the test. There two main actions; a search (by name and date of birth),
+and a direct lookup (by system number). Each of these actions has two different
+forms; a _probably_ successful form where the parameters are randomly selected
+from the corresponding payload file, and a _probably_ not successful form where
+parameters are randomly generated (see `./src/randomisor.js`).
```
I wonder if it's worth mentioning why this is?

i.e.
1. Random searches that are likely to return no results.
2. Less random searches that are likely to return a result but run the risk of getting cached by the database reducing the validity of the test.

daniel-ac-martin replied:
Ah you do mention this below...

daniel-ac-martin commented on README.md:
```diff
@@ -1,2 +1,49 @@
+Normally a mix of these scenarios is run (weighted in favour of the generated
+kind), to force the DB to perform searches (rather than simply read from
+caches). Therefore `404` is considered to be a successful HTTP response code
+(in addition to `200`, obviously). This is quite expected, as there will likely
+always be LEV users searching for records that don't exist.
```
This isn't quite true. I think the vast majority of searches return results as otherwise the user would be 'fishing' or going on bad information.

RobinKnipe replied:
I've reworded this part.

daniel-ac-martin commented:
Could we mention some basic stuff around dependencies and the need to run `npm install` beforehand?

daniel-ac-martin commented:
Is it worth writing something about the TCP (and therefore TLS) connection pools vs HTTP requests?

daniel-ac-martin commented:
Do we need stuff around secrets management?

RobinKnipe committed:
Minor content changes to reflect PR comments.

RobinKnipe committed:
Add an example config template file

...to show the available env vars.

RobinKnipe committed:
Assume tests should run against QA env by default

...otherwise the `TARGET_API` env var can be set in the config file to
switch to an alternative test target.

RobinKnipe committed:
Check and alert for no Artillery.io installation

RobinKnipe commented:
> Could we mention some basic stuff around dependencies and the need to run `npm install` beforehand?

Added a check for this in `entrypoint.sh`, which will alert the user. Fixed

RobinKnipe committed:
A note about secret management and the config file

RobinKnipe commented:
> Do we need stuff around secrets management?

Added a note about secret management and the "config file". Fixed

RobinKnipe committed:
A note on connection pooling and high CPU warnings

RobinKnipe commented:
> Is it worth writing something about the TCP (and therefore TLS) connection pools vs HTTP requests?

I've added a note about this and the problems of high CPU warnings while testing. Fixed

daniel-ac-martin commented on config-example.template.conf:
```diff
@@ -0,0 +1,24 @@
+
+# Certificate and key files for mutual TLS
+CLIENT_CERT="./config/client.crt"
+CLIENT_KEY="./config/private.key"
```
Should we add something to `.gitignore`?
```
*.key
```
?

RobinKnipe replied:
That whole folder `./config` is already ignored

daniel-ac-martin replied:
Okay.

daniel-ac-martin commented on entrypoint.sh:
```diff
@@ -1,9 +1,17 @@
 #!/bin/bash
 
+export TARGET_API="https://api.qa.notprod.lev.homeoffice.gov.uk"
```
Should this be hard-coded here?
Shouldn't it come from the config file?

RobinKnipe replied:
That's just a default, the config files for other envs set an explicit target

daniel-ac-martin replied:
It's a bit odd to set that but not anything else.

Would it make more sense to remove it and instead do:
```bash
set -euo pipefail
```

daniel-ac-martin replied:
I suppose it's not a big deal. Have a think. I'm happy to merge with no changes here.

RobinKnipe replied:
I think because most/all of the rest of the config values are "sensitive" this is the only one where it makes sense to set a default. I think I'm inclined to leave it as is...

daniel-ac-martin commented on entrypoint.sh:
```diff
@@ -1,9 +1,17 @@
   echo "Error: expected to be called with a configuration file"
   exit 1
 fi
+if ! artillery -V &> /dev/null ; then
+  echo "Error: Artillery is not installed, either run:
+  - npm i # to install it locally, or
+  - npm install -g artillery # for global installation"
+  exit 1
+fi
```
Does this not require that it be install globally?
Is it worth adding `./node_modules/.bin` to the path?

RobinKnipe replied:
Yes, good point, sometimes I forget that I already have this as part of my `.bashrc`. Refactored `./entrypoint.sh`

RobinKnipe committed:
Add local node binaries/executables to the PATH

...so that locally installed versions of Artillery can be used,
not just  global ones.

daniel-ac-martin commented on entrypoint.sh:
```diff
@@ -1,9 +1,18 @@
 #!/bin/bash
 
+export TARGET_API="https://api.qa.notprod.lev.homeoffice.gov.uk"
+export PATH="$PATH:./node_modules/.bin"
```
I'm not sure this wants `export` does it?

The scope should be limited to this script.
